stdlib/AllocProfile/src/AllocProfile.jl

```diff
@@ -0,0 +1,187 @@
+module AllocProfile
+
+using Base.StackTraces: StackTrace, StackFrame, lookup
+using Base: InterpreterIP
+
+# raw results
+
+# matches RawBacktrace on the C side
+struct RawBacktrace
+    data::Ptr{Csize_t} # in C: *jl_bt_element_t
+    size::Csize_t
+end
+
+# matches RawAlloc on the C side
+struct RawAlloc
+    type::Ptr{Type}
+    backtrace::RawBacktrace
+    size::Csize_t
+end
+
+struct FreeInfo
+    type::Ptr{Type}
+    count::UInt
+end
+
+# matches RawAllocResults on the C side
+struct RawAllocResults
+    allocs::Ptr{RawAlloc}
+    num_allocs::Csize_t
+
+    frees::Ptr{FreeInfo}
+    num_frees::Csize_t
+end
+
+"""
+    AllocProfile.@profile [skip_every=10_000] ex
```

Comment:
I just realized the name here is a bit misleading/confusing:
We're not skipping every 10,000th element, we're skipping 10,000 elements after every sample. But as it's currently written, it sounds to me like "skip one every 10,000".

Maybe switch this to `sample_every=10_000`? I think that's unambiguous.

Reply:
Or we could switch to a rate, like `sample_rate=0.0001`? It seems like the current approach requires a tiny bit of paying close attention to flip the big numbers into smaller profiles in my head.

But on the other hand, the current format matches the `delay=0.01` field, in that bigger delays also mean smaller profiles.. So maybe this is actually best how you already have it, since it's more consistent. 🤔

Reply:
Oh, blech, but also `sample_every=0` doesn't make sense... So scratch that.

`skip_between=n` maybe? 

Also another problem with `sample_rate=0.001` is that then you need to convert this float into an integer anyway to do the skipping, which is weird.. Unless you actually do want to randomly sample via rand() instead of picking every N allocations, which might actually be better to avoid you missing any cyclical patterns if your skip rate and allocation patterns are in-phase?

For example, this would be a bummer:
```julia
Allocs.@profile skip_every=2 begin   # profile every third alloc (i think?)
    for _ in 1:1000
        a,b,c = Int[], Dict(), Float64[]
        # ...
    end
end
```
You'd only see the `Int[]` allocations and miss the other ones 😬 

So after thinking about that. I think I vote strongly in favor of taking a sample rate, and actually doing uniform random sampling with the provided sample rate. `1.0` means sample everything, `0.0` means sample nothing, and the default could be `0.0001` (sample every 1/10,000).

Reply:
uniform sampling is done